src/App.tsx

```diff
@@ -1,24 +1,206 @@
+
+    try {
+      setLoadingTodosIds(prev => [...prev, ...completedTodosIds]);
+      await Promise.all(completedTodosIds.map(id => deleteTodo(id)));
+      setTodos(prev => prev.filter(({ id }) => (
+        !completedTodosIds.includes(id)
+      )));
+    } catch {
+      setHasError(true);
+      setErrorType(AppError.Delete);
+    } finally {
+      setLoadingTodosIds(prev => (
+        prev.filter(id => !completedTodosIds.includes(id))
+      ));
+    }
+  };
+
+  const updateTodo = async (
+    todoId: number,
+    data: Partial<Omit<Todo, 'id'>>,
+  ) => {
+    try {
+      setLoadingTodosIds(prev => [...prev, todoId]);
+      await patchTodo(todoId, data);
+      setTodos(prev => prev.map((todo) => {
+        if (todo.id === todoId) {
+          return { ...todo, ...data };
+        }
+
+        return todo;
+      }));
+    } catch {
+      setHasError(true);
+      setErrorType(AppError.Patch);
+    } finally {
+      setLoadingTodosIds(prev => prev.filter(id => id !== todoId));
+    }
+  };
+
+  const toggleAllTodos = async () => {
+    let updatingTodosIds: number[];
+
+    try {
+      if (activeTodosCount) {
+        updatingTodosIds = todos
+          .filter(({ completed }) => !completed)
+          .map(({ id }) => id);
+        setLoadingTodosIds(prev => [...prev, ...updatingTodosIds]);
+        await Promise.all(updatingTodosIds.map((
+          id => (updateTodo(id, { completed: true }))
+        )));
+      } else {
+        updatingTodosIds = todos
+          .filter(({ completed }) => completed)
+          .map(({ id }) => id);
+        setLoadingTodosIds(prev => [...prev, ...updatingTodosIds]);
+        await Promise.all(updatingTodosIds.map((
+          id => (updateTodo(id, { completed: false }))
+        )));
+      }
+    } catch {
+      setHasError(true);
+      setErrorType(AppError.Patch);
+    } finally {
+      setLoadingTodosIds(prev => prev.filter((
+        id => !updatingTodosIds.includes(id)
+      )));
+    }
+  };
+
   if (!USER_ID) {
     return <UserWarning />;
   }
 
   return (
-    <section className="section container">
-      <p className="title is-4">
-        Copy all you need from the prev task:
-        <br />
-        <a href="https://github.com/mate-academy/react_todo-app-add-and-delete#react-todo-app-add-and-delete">React Todo App - Add and Delete</a>
-      </p>
-
-      <p className="subtitle">Styles are already copied</p>
-    </section>
+    <div className="todoapp">
+      <h1 className="todoapp__title">todos</h1>
+
+      <div className="todoapp__content">
+        <Header
+          hasTodos={!!todos.length}
+          toggleAllTodos={toggleAllTodos}
+          hasActiveTodos={!!activeTodosCount}
+          addTodo={addTodo}
+        />
+
+        <TodoList
+          onUpdate={updateTodo}
+          onDelete={removeTodo}
+          todos={visibleTodos}
+          loadingTodosIds={loadingTodosIds}
+        />
+
+        {tempTodo && <TodoItem todo={tempTodo} />}
```

Comment:
i'm not sure but for more readable structure of project better to do that in todolist component

Reply:
I probably agree. Just want to hear what mentor said